### ISSUE TITLE
New lint to make parameters/variables/fields non-nullable if ! is used unconditionally for them.

### ISSUE THREAD

**Issue by polina-c** [dart-lang/linter#3266]
For example, for the code

```
void hello(String? name){
  print('Hello, ${name!.trim()}!');
}
```

the lint would suggest:

```
void hello(String name){
  print('Hello, ${name.trim()}!');
}
```

**Comment by bwilkerson** [comment]
At the very least, the lint would need to not flag parameters in methods that override another.

I have to wonder whether there are other cases that would need to be taken into account.

**Comment by srawlins** [comment]
Another would be the case of the function being passed as a tear-off to some location that requires function with a nullable parameter.

```dart
typedef Fn = void Function(int?);
void f(int? i) {
  i!;
}
void g(Fn) {}
void main() {
  g(f);
}
```

**Comment by jacob314** [comment]
There are reasonable cases where a developer might want to make a parameter to a method nullable but fail any time it is not null so this lint may have too many false positives. For example, methods in a test framework could frequently have this pattern.

However, right now the null safety migration tool may generate cases like this fairly frequently so it would be nice to give users a hint that this pattern usually indicates code they could have written more cleanly.  Perhaps this is a good case for a suggestion level lint that only shows up in IDEs and code reviews when you modified the method using this pattern. It would be great if while reviewing a null safety migration CL I saw each case like this so I could make sure it was intentional rather than an indication of an API that had been migrated poorly.
